File: Best_Time_to_Buy_and_Sell_Stock_IV.java
/**
 * Problem Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/
 * 
 * Solution: 
 *      http://blog.csdn.net/linhuanmars/article/details/23236995
 *      http://www.cnblogs.com/grandyang/p/4295761.html
 */

public class Solution {
    public int maxProfit(int k, int[] prices) {
        int n = prices.length;
        if(n < 2 || k <= 0) return 0;
        
    	if (k >= n)
    		return solveMaxProfit(prices);
 
        int[][] local = new int[n][k+1];
        int[][] global = new int[n][k+1];
        
        for(int i=1; i<n; i++) {
            int diff = prices[i] - prices[i-1];
            for(int j=1; j<=k; j++) {
                local[i][j] = Math.max(global[i-1][j-1] + Math.max(0, diff), local[i-1][j] + diff);
                global[i][j] = Math.max(local[i][j], global[i-1][j]);
            }
        }
        return global[n-1][k];
    }
    
    private int solveMaxProfit(int[] prices) {
        int res = 0;
        for (int i = 1; i < prices.length; ++i) {
            if (prices[i] - prices[i - 1] > 0) {
                res += prices[i] - prices[i - 1];
            }
        }
        return res;
    }
}

File: Swift/best-time-to-buy-and-sell-stock-iv.swift
/**
 * https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/
 * 
 * 
 */ 
// Date: Mon Jun  1 14:33:28 PDT 2020
class Solution {
    func maxProfit(_ k: Int, _ prices: [Int]) -> Int {
        let n = prices.count
        guard n > 1, k > 0 else { return 0 }
        if k >= (n / 2) {
            var profit = 0
            for index in 1 ..< n {
                profit += max(0, prices[index] - prices[index - 1])
            }
            return profit
        }
        var dp = Array(repeating: Array(repeating: 0, count: n + 1), count: k + 1)
        for kk in 1 ... k {
            var tmpMin = prices[0] - dp[kk - 1][0]
            for sell in 2 ... n {
                tmpMin = min(tmpMin, prices[sell - 1] - dp[kk - 1][sell - 1])
                dp[kk][sell] = max(dp[kk][sell - 1], prices[sell - 1] - tmpMin)
            }
        }
        // print("\(dp)")
        return dp[k][n]
    }
}
